x/voting/keeper/delegate_mint_burn.go

```diff
@@ -0,0 +1,145 @@
+	if err != nil {
+		return err
+	}
+
+	err = k.delegationHandler(ctx, msg, ballotBefore, ballotAfter)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (k Keeper) delegationHandler(ctx sdk.Context, msg *types.MsgCreateVote, ballotBefore int64, ballotAfter int64) error {
+	ballotDiff := ballotAfter - ballotBefore
+
+	if ballotDiff > 0 { // delegate
+		ballotDiffAdjusted := ballotAfter - Max(ballotBefore, 0)
+		if ballotDiffAdjusted > 0 {
+			err := k.delegate(ctx, msg, ballotDiffAdjusted)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	// un-delegate
+	if ballotBefore > 0 {
+		ballotDiffAdjusted := ballotBefore - Max(ballotAfter, 0)
+		err := k.undelegate(ctx, msg, ballotDiffAdjusted)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (k Keeper) delegate(ctx sdk.Context, msg *types.MsgCreateVote, amount int64) error {
+	// mint stake token to voting module account
+	coinsToMint := sdk.Coins{sdk.NewInt64Coin("stake", amount)}
+	err := k.bankKeeper.MintCoins(ctx, types.ModuleName, coinsToMint)
+	if err != nil {
+		return err
+	}
+
+	// delegate minted stake tokens to receiver
+	votingModuleAcct := sdk.AccAddress(crypto.AddressHash([]byte(types.ModuleName)))
+
+	receiverAccAddress, _ := sdk.AccAddressFromBech32(msg.Receiver)
+	receiverValAddr, err := sdk.ValAddressFromHex(hex.EncodeToString(receiverAccAddress.Bytes()))
+	if err != nil {
+		return err
+	}
+
+	recipientValidator, found := k.stakingKeeper.GetValidator(ctx, receiverValAddr)
+	if !found {
+		return sdkerrors.Wrapf(types.ErrReceiverIsNotAValidator, receiverValAddr.String())
+	}
+
+	_, err = k.stakingKeeper.Delegate(ctx, votingModuleAcct, sdk.NewInt(amount), stakingtypes.Unbonded, recipientValidator, true)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (k Keeper) undelegate(ctx sdk.Context, msg *types.MsgCreateVote, amount int64) error {
+	// undelegate the stake tokens
+	votingModuleAcct := sdk.AccAddress(crypto.AddressHash([]byte(types.ModuleName)))
+	receiverAccAddress, _ := sdk.AccAddressFromBech32(msg.Receiver)
+	receiverValAddr, err := sdk.ValAddressFromHex(hex.EncodeToString(receiverAccAddress.Bytes()))
+	if err != nil {
+		return err
+	}
+
+	recipientValidator, found := k.stakingKeeper.GetValidator(ctx, receiverValAddr)
+	if !found {
+		return sdkerrors.Wrapf(types.ErrReceiverIsNotAValidator, receiverValAddr.String())
+	}
+
+	sharesToUnDelegate, err := recipientValidator.SharesFromTokens(sdk.NewInt(amount))
+	if err != nil {
+		return err
+	}
+
+	_, err = k.stakingKeeper.Undelegate(ctx, votingModuleAcct, receiverValAddr, sharesToUnDelegate)
+	if err != nil {
+		return err
+	}
+
+	// burn the stake tokens from voting module
+	//voteCountEquivalentStake := sdk.Coins{sdk.NewInt64Coin("stake", voteCountAbs)}
+	//err = k.bankKeeper.BurnCoins(ctx, types.ModuleName, voteCountEquivalentStake)
+	//if err != nil {
+	//	return err
+	//}
```

Comment:
Nitpick: Let's remove this as its not required anymore

Reply:
Done